src/actions/write/initiateFoundationAction.ts

```diff
@@ -0,0 +1,165 @@
+import {
+  ALLOWED_ACTIVE_TIERS,
+  DEFAULT_INVALID_ID_TIER_MESSAGE,
+  DEFAULT_INVALID_TIER_MESSAGE,
+  FOUNDATION_ACTION_ACTIVE_STATUS,
+  FOUNDATION_ACTION_PASSED_STATUS,
+  MAX_FOUNDATION_ACTION_PERIOD,
+  MAX_NAME_LENGTH,
+  MAX_NOTE_LENGTH,
+} from '../../constants';
+import {
+  ActiveTier,
+  ContractResult,
+  FoundationAction,
+  IOState,
+  PstAction,
+  ServiceTier,
+} from '../../types';
+import { isValidArweaveBase64URL } from '../../utilities';
+
+declare const ContractError;
+declare const SmartWeave: any;
+
+// Proposes a foundation action
+export const initiateFoundationAction = async (
+  state: IOState,
+  { caller, input: { type, note, value } }: PstAction,
+): Promise<ContractResult> => {
+  const foundation = state.foundation;
+
+  // The caller must be in the foundation, or else this action cannot be initiated
+  if (!foundation.addresses.includes(caller)) {
+    throw new ContractError(
+      `${caller} Caller needs to be in the foundation wallet list.`,
+    );
+  }
+
+  if (typeof note !== 'string' || note.length > MAX_NOTE_LENGTH) {
+    throw new ContractError('Note format not recognized.');
+  }
+
+  if (type === 'addAddress' && typeof value === 'string') {
+    if (!isValidArweaveBase64URL(value)) {
+      throw new ContractError(
+        'The target of this action is an invalid Arweave address?"',
+      );
+    }
+    if (foundation.addresses.includes(value)) {
+      throw new ContractError(
+        'Target is already added as a Foundation address',
+      );
+    }
+  } else if (type === 'removeAddress' && typeof value === 'string') {
+    if (!foundation.addresses.includes(value)) {
+      throw new ContractError(
+        'Target is not in the list of Foundation addresses',
+      );
+    }
+  } else if (type === 'setMinSignatures' && typeof value === 'number') {
+    if (
+      !Number.isInteger(value) ||
```

Comment:
(complement) 
Not sure if this was the intent but

```js
> typeof NaN
'number'
```

is a nasty one

but

```js
> Number.isInteger(NaN)
false
```

checks for it 👍